mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAnnotationMapper.java

```diff
@@ -0,0 +1,159 @@
+import org.apache.james.mailbox.cassandra.CassandraId;
+import org.apache.james.mailbox.cassandra.table.CassandraAnnotationTable;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.model.MailboxAnnotation;
+import org.apache.james.mailbox.model.MailboxAnnotation.AnnotationValue;
+import org.apache.james.mailbox.model.MailboxAnnotation.MailboxAnnotationCommand;
+import org.apache.james.mailbox.model.MailboxAnnotation.MailboxAnnotationEntryKey;
+import org.apache.james.mailbox.model.MailboxAnnotation.MailboxAnnotationEntryValue;
+import org.apache.james.mailbox.model.SimpleMailboxAnnotation;
+import org.apache.james.mailbox.model.SimpleMailboxAnnotation.SimpleMailboxAnnotationEntryKey;
+import org.apache.james.mailbox.model.SimpleMailboxAnnotation.SimpleMailboxAnnotationEntryValue;
+import org.apache.james.mailbox.store.mail.model.Mailbox;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.datastax.driver.core.ResultSet;
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.Session;
+import com.google.common.base.Preconditions;
+
+public class CassandraAnnotationMapper {
+
+    @FunctionalInterface
+    public interface CodeInjector {
+        void inject();
+    }
+
+    private final Mailbox<CassandraId> mailbox;
+    private final Session session;
+    private final int maxRetry;
+    private final CodeInjector codeInjector;
+
+    private static final Logger LOG = LoggerFactory.getLogger(CassandraAnnotationMapper.class);
+
+    public CassandraAnnotationMapper(Mailbox<CassandraId> mailbox, Session session, int maxRetry) {
+        this(mailbox, session, maxRetry, () -> {
+        });
+    }
+
+    public CassandraAnnotationMapper(Mailbox<CassandraId> mailbox, Session session, int maxRetry,
+            CodeInjector codeInjector) {
+        Preconditions.checkArgument(maxRetry > 0);
+        this.mailbox = mailbox;
+        this.session = session;
+        this.maxRetry = maxRetry;
+        this.codeInjector = codeInjector;
+    }
+
+    public MailboxAnnotation getAnnotation() {
+        ResultSet resultSet = getStoredAnnotationRow();
+        if (resultSet.isExhausted()) {
+            return SimpleMailboxAnnotation.EMPTY;
+        }
+        Map<MailboxAnnotationEntryKey, MailboxAnnotationEntryValue> annos = CassandraUtils.convertToStream(resultSet)
+                .map(row -> new AbstractMap.SimpleEntry<>(buildKey(row.getString(CassandraAnnotationTable.KEY)),
+                        buildValue(row.getString(CassandraAnnotationTable.VALUE))))
+                .collect(Collectors.toMap(AbstractMap.SimpleEntry::getKey, AbstractMap.SimpleEntry::getValue));
+
+        return new SimpleMailboxAnnotation(annos);
+    }
+
+    public void updateAnnotation(MailboxAnnotationCommand command) throws MailboxException {
+        try {
+            new FunctionRunnerWithRetry(maxRetry).execute(() -> {codeInjector.inject();
```

Comment:
If you go managing each annotation separately, and not as a wall, you can get ride of function retry, and lightweight transactions and get a safe, and yet way faster code.